agents-for-bedrock/features-examples/08-create-agent-with-guardrails/08-create-agent-with-guardrails.ipynb

```diff
@@ -0,0 +1,1956 @@
+    "    except Exception as e:\n",
+    "        raise Exception(\"unexpected event.\", e)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2bd6097d",
+   "metadata": {
+    "pycharm": {
+     "name": "#%% md\n"
+    }
+   },
+   "source": [
+    "##### Invoke Agent to query Knowledge Base\n",
+    "Let's now use our support `invokeAgent` function to query our Knowledge Base with the Agent"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "bb134630-0c01-4ec7-b761-1f1a1eca4957",
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    },
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "import uuid\n",
+    "session_id:str = str(uuid.uuid1())\n",
+    "query = \"How can I activate my debit card?\"\n",
+    "response = invokeAgent(query, session_id)\n",
+    "print(response)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e59b3614",
+   "metadata": {
+    "pycharm": {
+     "name": "#%% md\n"
+    }
+   },
+   "source": [
+    "##### Invoke Agent to execute function from Action Group\n",
+    "Now let's test our Action Group functionality and create a new reservation"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1e0201c0-abb3-4ad5-a7e7-1750d14a2717",
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    },
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "query = \"Hi, I am Anna. I want to create a banking appointment for 8pm on the 5th of May 2024.\"\n",
+    "response = invokeAgent(query, session_id)\n",
+    "print(response)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c7043ba4",
+   "metadata": {
+    "pycharm": {
+     "name": "#%% md\n"
+    }
+   },
+   "source": [
+    "##### Invoke Agent with prompt attribute\n",
+    "\n",
+    "Great! We've used our agent to do the first reservation. However, often when booking tables in restaurants we are already logged in to systesm that know our names. How great would it be if our agent would know it as well?\n",
+    "\n",
+    "To do so, we can use the session context to provide some attributes to our prompt. In this case we will provide it directly to the prompt using the [`promptSessionAttributes`](https://docs.aws.amazon.com/bedrock/latest/userguide/agents-session-state.html) parameter. Let's also start a new session id so that our agent does not memorize our name."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "139c210f",
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "session_id:str = str(uuid.uuid1())\n",
+    "query = \"Should I invest in bitcoin?\"\n",
+    "session_state = {\n",
+    "    \"promptSessionAttributes\": {\n",
+    "        \"name\": \"John\"\n",
+    "    }\n",
+    "}\n",
+    "response = invokeAgent(query, session_id, session_state=session_state)\n",
```

Comment:
getting the following error:

`Exception: ('unexpected event.', EventStreamError('An error occurred (accessDeniedException) when calling the InvokeAgent operation: User: arn:aws:sts::864675421752:assumed-role/AmazonBedrockExecutionRoleForAgents_banking-assistant1q/BedrockAgentRetrieveSession-FLYBLH8XV1 is not authorized to perform: bedrock:Retrieve on resource: arn:aws:bedrock:us-west-2:864675421752:knowledge-base/FLYBLH8XV1 because no identity-based policy allows the bedrock:Retrieve action'))`